add interface abstraction and queue mapping

cjdoucette opened the pull request:
The point of this pull request is to add more networking configuration that will help the functional blocks, and also to add the beginnings of the BP block.

The operator needs to be able to specify which ports are the Gatekeeper front and back, so the second patch sets up the ability to specify which ports belong to which interface and creates a single identifier for receiving/transmitting.

The BP implementation is only a first effort, as we must answer some key questions before continuing. For example, if ntuple filter is not flexible, how will IPv4 rules be added? Additionally, will this block be able to listen directly on a queue, or will it have to wait for packets to be distributed to it in software? The next step is answering these questions for IPv4 (which will help other blocks), and then adding support for IPv6.

AltraMayor commented:
We should postpone the BP patch until we have something concrete there. For now, let's focus on the blocks that can be implemented on our NIC. Drop the BP commit, and add a component that can be implemented instead.

cjdoucette commented:
OK, I removed the BP commit for now. I will work on the LLS block instead.

mengxiang0811 commented on lib/net.c:
```diff
@@ -50,32 +51,231 @@ find_num_numa_nodes(void)
+	uint8_t num_succ_ports = 0;
+	uint8_t num_slaves_added = 0;
+
+	iface->ports = malloc(sizeof(*iface->ports) * iface->num_ports);
```
Should we use ``rte_malloc()`` or ``rte_zmalloc()`` instead of ``malloc()`` or ``calloc()`` for all the cases? These functions may perform better in NUMA systems.

mengxiang0811 commented on lib/net.c:
```diff
@@ -50,32 +51,231 @@ find_num_numa_nodes(void)
+
+	iface->ports = malloc(sizeof(*iface->ports) * iface->num_ports);
+	if (iface->ports == NULL) {
+		RTE_LOG(ERR, EAL, "%s: Out of memory for %s ports\n",
```
If so, we also need to change ``EAL`` to ``MALLOC``.

cjdoucette replied:
Agreed.

mengxiang0811 commented:
**gatekeeper: reduce logging to one logtype**
This commit solves the issue with limited number of custom log types. I think this commit is ready to be merged.

**net: add interface abstraction**

cjdoucette commented on lib/net.c:
```diff
@@ -50,32 +51,231 @@ find_num_numa_nodes(void)
+	if (iface->num_ports == 1)
+		return;
+
+	for (i = 0; i < nb_slave_ports; i--) {
```
Thank you -- yes.

cjdoucette commented:
I've updated the pull request with a few patches that make minor changes to things like allocation and how we get the queue identifiers.

I've also added patches that fill in the LLS block by adding ARP and LACP support.

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -40,4 +42,14 @@ int gatekeeper_init_network(struct net_config *net_conf);
+static inline uint16_t
+get_queue_id(void)
+{
+	return (uint16_t)rte_lcore_id();
```
This patch is assuming that the id of the lcores are sequential and starting at zero, what's not always true. There are parameters of DPDK to force the use of the highest lcores for example.

A solution for this problem is for `get_queue_id()` to use an array whose index is the lcore id, and its values the queue id. Of course this array will need to be initialized, and `gatekeeper_init_network()` can easily do so.

AltraMayor replied:
I've confused `lcore_id` with lcore index. Could you try to call Gatekeeper skipping some cores to see if the current code will work fine?

cjdoucette replied:
The lcore IDs indeed do *not* always start from 0. They are used exactly as they are identified with any command line parameters.

Therefore, I've added the array you suggested, which allows us to use queue numbers starting from 0 with any set of lcore IDs.

AltraMayor commented on lib/net.c:
```diff
@@ -48,31 +49,218 @@ find_num_numa_nodes(void)
+close_iface(struct gatekeeper_if *iface)
+{
+	close_iface_id(iface, iface->num_ports);
+	close_iface_ports(iface, iface->num_ports);
```
`close_iface_id()` and `close_iface_ports()` should drop their second parameters and use `iface->num_ports`. They already receive `iface`.

cjdoucette replied:
During initialization, if initializing some of the ports fails, `close_iface_id` and `close_iface_ports` are called with a second parameter that designates how many successful port openings there were. So I intended to re-use that code and call it with all the ports here.

AltraMayor replied:
Fine.

AltraMayor commented on lua/net.lua:
```diff
@@ -1,10 +1,14 @@
 function M.setup_block()
 
+	local front_ports = {"0000:83:00.0"}
+	local back_ports = {"0000:85:00.0"}
```
We need to use the name that the Linux kernel uses to name network interfaces. Using PCI addresses would be a big surprise to network administrators in a negative way. A pointer to start is to look at the code of command `xip`. `xip` has to interpret interface names for some cases. It should be possible to map interface names to PCI addresses.

AltraMayor replied:
Isn't the ids that DPDK uses to interfaces the same that the kernel uses? If so, all you need to do is to get the id the kernel uses; that's exactly what `xip` does.

cjdoucette replied:
I tried using the standard library call for this (`if_nametoindex`), but it returns an error when run with interface names that have been disconnected from the kernel driver. The scripts for binding and unbinding interfaces are also unable to work with interface names when they are not bound to the kernel.

For now, I amended the DPDK binding/unbinding scripts so that whenever they see the name of an interface, they append it (and its PCI address) to the file `lua/gatekeeper_ifs.txt` if it's not already there. You can see this change here: https://github.com/cjdoucette/dpdk/commit/ca9e4d03b720bb958d9c35648cb85117b8c44132

It's not perfect -- for example, if multiple users are using the system and only one executes the bind commands, only she will have the interface names in her repository. So this is the idea but we should think about improvements.

cjdoucette replied:
That sounds much cleaner, thanks. Any preference on where this new file should go? For example, we could place it in `/etc/if_map.lua` so that all users can access it or could put it in the gatekeeper repository itself.

AltraMayor replied:
The `setup.sh` should write it at the folder "lua" of the repository. And our Lua code can assume the existence of the file.

AltraMayor commented on lua/net.lua:
```diff
@@ -13,6 +17,20 @@ function M.setup_block()
 
+	conf.front.num_ports = #front_ports
+	conf.front.name = "front"
+	conf.front.pci_addrs = ffi.new("const char *[" .. #front_ports .. "]")
```
You'll need a C function to properly work here. The current code is wrong in a subtle way: the strings held in `front_ports` and `back_ports` are allocated by Lua, and Lua will release those string once it is done setting Gatekeeper up. The problem will come up when `conf.*.pci_addrs` is accessed later on.

cjdoucette replied:
Maybe there's an easier way than the solution I came up with, but I ran into the issue that in order to pass the strings from Lua to C, they have to be declared as `const char **`, so you can't later re-use that part of `struct net_config` to do dynamic allocation and store the strings permanently for C. So I ended up using `struct net_config` to both pass constant parameters from Lua (with a note saying not to use them) and also hold the dynamically allocated versions. from

cjdoucette replied:
This is what I did in the rebase that I pushed -- I was just commenting here that I couldn't come up with a better solution than having two representations of these strings in `struct net_config` -- one ephemerally for Lua to pass to C, and one permanently for C to use throughout the life of the program.

AltraMayor commented on include/gatekeeper_lls.h:
```diff
@@ -19,6 +19,12 @@
+	unsigned int	lcore_id;
+};
+
+struct lls_config *alloc_lls_conf(void);
```
Shouldn't there be only one instance of LLS? If so, this function should be called `get_lls_conf()` and only return that address.

cjdoucette replied:
Yes, point taken. Thanks!

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -38,11 +37,20 @@ struct gatekeeper_if {
 	const char	*name;
 
+	/* Timeout for cache entries (in seconds) for Link Layer Support. */
+	uint32_t	arp_cache_timeout;
```
Renaming this field to `arp_cache_timeout_sec` makes it clear that the unit is seconds everywhere it's used.

cjdoucette replied:
Done.

AltraMayor commented on lls/main.c:
```diff
@@ -16,11 +16,206 @@
+	arp_cache_destroy();
+	modify_arp_filter(net_conf->back.id, (uint16_t)rte_lcore_id(),
+		RTE_ETH_FILTER_DELETE);
+	modify_arp_filter(net_conf->front.id, (uint16_t)rte_lcore_id(),
```
`get_queue_id()` should be used here, not `rte_lcore_id()` in the two cases above.

AltraMayor commented on lls/main.c:
```diff
@@ -16,11 +16,206 @@
+		GATEKEEPER_MAX_PKT_BURST);
+	uint16_t i;
+
+	/* XXX Is this likely or unlikely? */
```
It depends on the timeout of all machines, so we don't know it a priori. Just drop this comment.

AltraMayor commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+		     arp_hdr->arp_pln != sizeof(uint32_t)))
+		return -1;
+
+	/* XXX If sip is not in the same subnet as our IP address, drop. */
```
Change XXX to TODO, it is very important.

cjdoucette replied:
Changed. (Is this as simple as adding a prefix length to the config file and checking against it here?)

AltraMayor replied:
Yes, it is.

AltraMayor commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+
+	/* Check if we have this IP address. */
+	target_ip_be_32 = arp_hdr->arp_data.arp_tip;
+	if (iface->ipv4_addr != rte_be_to_cpu_32(target_ip_be_32))
```
What about broadcasts?

cjdoucette replied:
Thanks -- I added another test so that broadcasts in the form of GARP packets (two formats) can also pass through.

AltraMayor commented on lls/arp.h:
```diff
@@ -0,0 +1,55 @@
+
+#include "gatekeeper_lls.h"
+
+struct arp_cache_entry {
```
There is no field to count how many users of this entry exit. Two blocks may ask LLS to monitor a given IP address. If one block revokes the request, the entry shouldn't be removed until the second block also revokes its request.

AltraMayor commented on lls/main.c:
```diff
@@ -19,11 +19,15 @@
 #include "gatekeeper_net.h"
 #include "arp.h"
 
+/* For LACP to work best, RX burst size should be at least 2 * num slaves. */
```
Could you add the why as well?

cjdoucette replied:
Sure -- added.

AltraMayor commented on lls/main.c:
```diff
@@ -170,18 +174,43 @@ lls_proc(void *arg)
 		/* Process any ARP updates. */
 		arp_cache_writer();
+
+		/* Send gratuitous ARP to fulfill LACP requirement. */
```
If LACP is enabled on both interfaces, don't you need to transmit a packet on the second interface as well?

mengxiang0811 commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+	}
+
+	arp_cache_entry_pool = rte_mempool_create("arp_cache_entry_pool",
+		ARP_CACHE_ENTRIES / 2 - 1, sizeof(struct arp_cache_entry),
```
Should the size of the memory pool for entries be at least greater than ``ARP_CACHE_ENTRIES``? Since all the entries in the array should be from the memory pool.

cjdoucette replied:
I think the name was misleading here -- the size of the memory pool is the size of the *cache* of the memory pool, which should be smaller than `ARP_CACHE_ENTRIES`. I changed the name of the constant to emphasize that it's the cache of packet buffers for the ARP cache.

mengxiang0811 replied:
Thanks Cody, the name for the cache size is much clearer. For the second parameter in ``rte_mempool_create()``, it represents the number of elements in the mempool. Should it be ``ARP_CACHE_ENTRIES`` or ``ARP_CACHE_ENTRIES / 2 - 1``?

cjdoucette replied:
This should be fixed with the latest rebase.

mengxiang0811 commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+	arp_cache_entry_pool = rte_mempool_create("arp_cache_entry_pool",
+		ARP_CACHE_ENTRIES / 2 - 1, sizeof(struct arp_cache_entry),
+		ARP_CACHE_MEMPOOL_SIZE, 0, NULL, NULL, NULL, NULL,
+		rte_socket_id(), 0);
```
The ``socket id`` should be the id corresponding to the ``lcore`` of that functional block. Some other places may also need to be changed.

cjdoucette commented on lua/net.lua:
```diff
@@ -3,6 +3,21 @@ local ffi = require("ffi")
 
 local M = {}
 
+function ipv4_to_num(addr)
```
I ended up passing the IPv4 string to C to then call `inet_pton()`. Is that what you had in mind?

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+#include "arp.h"
+
+/*
+ *	ARP cache
```
Added.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+/* Length of time (in seconds) to wait between scans of the cache. */
+#define ARP_CACHE_SCAN_INTERVAL	10
+
+#define TIMER_RESOLUTION_CYCLES 20000000ULL /* around 10ms at 2 Ghz */
```
Removed.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,541 @@
+		}
+
+		printf("%s: ", ip_str);
+		printf("%02"PRIx8" %02"PRIx8" %02"PRIx8
```
Replaced.

cjdoucette commented:
I answered all comments in-line except one:

_The Lua interpreter has a couple macros to enable the host program (in our case, Gatekeeper) easily change how memory is allocated. Could you investigate if it's true for LuaJIT as well? If so, please add it to this commit._

I tinkered with this for a little while, but kept getting memory errors. It seems that LuaJIT does not support this out-of-box for 64-bit environments. I found this which seems to indicate it is perhaps possible with some advanced configuration of the address space:

https://gist.github.com/nddrylliog/8722197

At the moment, not sure where to begin with that (if it's possible with hugepages), and not sure it's worth it.

AltraMayor commented:
Based on what I read online, only LuaJIT 3.0 will properly address the issue of 64-bit memory allocation. The issue is the garbage collector of LuaJIT that doesn't scale. Let's leave LuaJIT as it is for now.

AltraMayor commented:
Having said that, the patch **gatekeeper: replace calloc with rte_calloc** is read to be merged.

cjdoucette commented:
The changes I just pushed are for "add interface abstraction" where I included a new C program to generate the mapping and "add ARP to LLS" where I changed the way the Lua script gets the IPv4 addresses in binary before passing them to initialize the network.

AltraMayor commented on lib/net.c:
```diff
@@ -88,11 +96,18 @@ gatekeeper_init_network(struct net_config *net_conf)
+
+	/* Assign queue identifier to each lcore. */
+	RTE_LCORE_FOREACH_SLAVE(lcore) {
+		net_conf->queues[lcore] = queue_id++;
```
Instead of pre-allocating a queue to each core, the array should be filled with an invalid queue identifier. When `lcore_to_queue()` is first called, it should allocate a queue, assign it to the array, and return it. The following time, it just returns what is in the array.

AltraMayor commented on include/gatekeeper_net.h:
```diff
@@ -32,6 +34,7 @@ struct net_config {
 	uint32_t		num_ports;
 	uint32_t		numa_nodes;
 	struct rte_mempool 	**gatekeeper_pktmbuf_pool;
+	uint16_t		queues[RTE_MAX_LCORE];
```
This allocation should be per interface, and per rx and tx. That is, we need four arrays.

The motivation for this is that not all blocks are going to need a single queue on both devices, and the type of those queues may be different. The GK-GT unit, for example, requires no queue from the front interface, and two queues from the back interface.

AltraMayor replied:
We need four arrays to accommodate queues per tx and rx and per interface.

The motivation for this is that the blocks don't have a uniform queue demand. The GK-GT unit for example requires a rx and tx queue from the back interface, but none from the front interface.

AltraMayor commented on lib/net.c:
```diff
@@ -64,6 +64,13 @@ get_net_conf(void)
 }
 
+uint16_t
+lcore_to_queue(unsigned int lcore)
```
This function needs to know if the queue is rx or tx, and from which interface.

AltraMayor replied:
This function needs two more parameters to know which kind of queue to map (i.e. tx or rx), and from which interface (i.e. front, and back).

AltraMayor commented on lua/net.lua:
```diff
@@ -7,8 +7,8 @@ function M.setup_block()
-	conf.num_rx_queues = 1
-	conf.num_tx_queues = 1
+	conf.num_rx_queues = 3
+	conf.num_tx_queues = 3
```
We need these parameters per interface.

AltraMayor replied:
We need these two parameters per interface.

AltraMayor commented on lib/net.c:
```diff
@@ -88,11 +96,18 @@ gatekeeper_init_network(struct net_config *net_conf)
+	RTE_ASSERT(net_conf->num_tx_queues < rte_lcore_count());
+
+	/* Assign queue identifier to each lcore. */
+	RTE_LCORE_FOREACH_SLAVE(lcore) {
```
The queue arrays should be initialized with an invalid queue identifier. When `lcore_to_queue()` sees an invalid number, it allocates a queue identifier, and returns that. The following calling will return the previously allocated identifier.

AltraMayor commented on lua/gatekeeperc.lua:
```diff
@@ -14,6 +14,7 @@ struct net_config {
 	uint32_t		num_ports;
 	uint32_t		numa_nodes;
 	struct rte_mempool 	**gatekeeper_pktmbuf_pool;
+	uint16_t		queues[128];
```
Can't we use the define here? I know that we'll still have to repeat the number in the define, but if the define is used again, we'll have avoided a duplicate. Moreover, knowing the identifier here helps to understand the limit.

cjdoucette commented:
The pull request now consists of four commits. To summarize:

The first two have been deemed ready to accept.

The third, **net: add interface abstraction**, has mostly already been seen. Please review again, but the main outstanding issues were:
- How to initialize strings and arrays of strings so that they can still be used after Lua frees its copies.
- Adding parentheses around the sizeof operator in generate_if_map.c.

The fourth, **net: improve queue configuration**, is mostly new code in response to the suggestions from yesterday morning. Permitting each functional block to choose whether it wants to be assigned an RX or TX queue on either interface makes this a little bit difficult, as explained in the commit message. In essence, the configuration must wait to allow each functional block to be assigned queue IDs before the network devices can be started. This break makes functions like `destroy_iface()` quite complicated.

The only alternative to this approach is to write each (lcore, interface, queue type) combination needed in the config files so that the configuration has all of this information _beforehand_. This is not very flexible.

I would be OK with removing this patch from the pull request for now, if needed, to get this pull request accepted and move on while we think about the design of the queue IDs.

mengxiang0811 commented on lua/net.lua:
```diff
@@ -1,17 +1,61 @@
+
+	-- Code below this point should not need to be changed.
+	conf.front.num_ports = #front_ports
+	conf.front.name = gatekeeperc.lua_init_iface_name("front")
```
After reviewing the code, I think all these code trying to do is to initialize the interface. Instead of exposing a set of C functions to finish the initialization, I wonder whether we can just expose one C function, e.g., ``init_iface("front", #front_ports, port_names)``, where ``port_names`` is the ``const char *`` array containing ``enp131s0f0``. Thus, we may not need the separate file `` generate_if_map.c``,  instead we can maintain the mapping function in lib/net.c. Therefore, this C function actually initializes the front interface.

cjdoucette replied:
Thanks, Qiaobin. Could you take a look at the most recent push? I put all of the interface initialization that Lua can do into one C function, which I think looks much better.

Although, I still think we need the separate file `generate_if_map.c`, since the point is to be able to run it while the interfaces are still bound to the kernel driver (during the setup stage). That way, the map is built before the interfaces are changed to the DPDK driver, because at that point we won't be able to easily get them. By the time `gatekeeper` is run, the names will be lost.

mengxiang0811 replied:
Thanks Cody for the clarification! The code looks simpler. Also, I have checked the APIs for getting the device information. Indeed, I found that the name mappings in DPDK are quite different from the kernel ones. So, I agree that it's necessary to obtain them before the network adapters are bound to DPDK.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+	rte_free(iface->ports);
+pci:
+	free_pci_addrs(iface);
+	rte_free(iface);
```
PCI addresses are not allocated in this function, so why would one need to release them here?

Moreover, the caller is passing a non-dynamically allocated address at `iface`, so this `rte_free()` will potentially lead to a segmentation fault.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+	uint8_t i;
+	for (i = 0; i < iface->num_ports; i++)
+		rte_free(iface->pci_addrs[i]);
+	rte_free(iface->pci_addrs);
```
Although not required, setting `iface->pci_addrs` to `NULL` after freeing the allocation it held would help to catch bugs rather quickly. For example, consider mistakenly calling `free_pci_addrs()` twice.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+				__func__, pci_addrs[i]);
+			for (j = 0; j < i; j++)
+				rte_free(iface->pci_addrs[j]);
+			rte_free(iface->pci_addrs);
```
Please set `iface->pci_addrs = NULL` after `rte_free()`. See justification in a comment below.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+	return 0;
+
+name:
+	rte_free(iface->name);
```
Please add `iface->name = NULL;` after `rte_free()`.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+lua_free_iface(struct gatekeeper_if *iface)
+{
+	free_pci_addrs(iface);
+	rte_free(iface->name);
```
Please NULL `iface->name` here too.

AltraMayor commented on lib/net.c:
```diff
@@ -47,32 +48,278 @@ find_num_numa_nodes(void)
+	close_iface_ports(iface, iface->num_ports);
+	rte_free(iface->ports);
+	free_pci_addrs(iface);
+	rte_free(iface->name);
```
NULL `iface->ports` and `iface->name`.

AltraMayor commented:
My comment on patch "net: add interface abstraction" are mostly touchups, but a comment is about a bug.

We do need a solution for allocating queues to functional blocks, but it's better to move commit "net: improve queue configuration" to its own pull request, so we can work on it in parallel to other pull requests.

AltraMayor commented on generate_if_map.c:
```diff
@@ -0,0 +1,100 @@
+		 */
+		if (iter->ifa_addr == NULL ||
+				iter->ifa_addr->sa_family != AF_PACKET ||
+				strncmp(iter->ifa_name, "lo", 2) == 0)
```
What's the intention here: (1) skipping the interface "lo", or (2) skipping all interfaces whose names start with "lo"? The current code is implementing the second case, but I think that the intention is the first one.
